File: design-patterns/cplusplus/src/creational/Singleton.cpp
#include "Singleton.h"

// Initialize static member variable
Singleton* Singleton::_instance = nullptr;

// Definition of getInstance() method
Singleton* Singleton::getInstance() {
    if (_instance == nullptr) {
        _instance = new Singleton();
    }
    return _instance;
}

// Definition of destructor
Singleton::~Singleton() {
    delete _instance;
}

File: design-patterns/singleton-pattern/java/Singleton.java
public class Singleton {

    // Private static instance variable
    private static Singleton instance;

    // Private constructor to prevent instantiation from outside
    private Singleton() {}

    // Static method to get the instance
    public static Singleton getInstance() {
        if (instance == null) {
            synchronized (Singleton.class) {
                if (instance == null) {
                    instance = new Singleton();
                }
            }
        }
        return instance;
    }

    // Method to perform some action
    public void performAction() {
        System.out.println("Singleton instance performing an action.");
    }
}
